### PR TITLE
Support Multi-Tenancy in Keycloak Authorization

### LINKED ISSUE #16973: add multi-tenancy support to keycloak-authorization
## Description

Add multi-tenancy support to keycloak-authorization so that it can be used together with the oidc-multi-tenancy.

## Implementation ideas

Configuration could be similar to the oidc-configuration:

```properties
# for default tenant
quarkus.keycloak.policy-enforcer.enable=true

# for specific tenant
quarkus.keycloak."tenant".policy-enforcer.enable=true
```

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #16973

This PR:

- moves `PolicyEnforcer` initialization code into a recorder
- structures the config similarly to `OidcConfig/OidcTenantConfig` with every tenant config also checked
- the policy check will get the required policy enforcer by checking `SecurityIdentity's` `tenant-id` attribute set by `quarkus-oidc`
- makes `keycloak.policy-enforcer.enable` build time only property
- I could not make the rest of the config runtime only since it is used to get a request body item during the build phase - but we can look at it later
- test added - but can't make it work, what I get is:
```
Expected: a string containing "Permission Resource Tenant"
  Actual: [{"claims":{"static-claim":["static-claim-tenant"],"request-uri":["/api-permission-tenant"],"request-method":["GET"]},"resourceId":"2ad0ef0e-421c-4537-b78a-cbd4270e8a51","resourceName":"Permission Resource","scopes":[]}]
```

I've done many various tries - and verified a `tenant` id is correctly selected. Pedro, have a look at the test please

### CODE REVIEW

**Comment by pedroigor** [comment]
@sberyozkin That is awesome! Thanks for it. Changes LGTM.

**Comment by sberyozkin** [comment]
Hey Pedro, @pedroigor, thanks, the only problem is the test :-) but it should be something minor

**Comment by sberyozkin** [comment]
@pedroigor, I have set `Permission Resource` in
```
quarkus.keycloak.tenant.policy-enforcer.paths.1.name=Permission Resource
quarkus.keycloak.tenant.policy-enforcer.paths.1.path=/api-permission-tenant
quarkus.keycloak.tenant.policy-enforcer.paths.1.claim-information-point.claims.static-claim=static-claim
```

because if I set to `Permission ResourceTenant`:

```
quarkus.keycloak.tenant.policy-enforcer.paths.1.name=Permission ResourceTenant
quarkus.keycloak.tenant.policy-enforcer.paths.1.path=/api-permission-tenant
quarkus.keycloak.tenant.policy-enforcer.paths.1.claim-information-point.claims.static-claim=static-claim
```

then it fails with `403` (and then by a trial and error I noticed that naming it `Permission Resource` avoids `403`but fails nonetheless).

`Permission Resource` name is also used in a similar default policy for `/api/permission` - so I'm not sure, when the above error is returned what permission is actually checked by Keycloak - the one named `Permission Resource` allocated for `api/permissions` or the one added with this PR, `Permission ResourceTenant`, allocated to handle `/api-permission-tenant` (this is why I also made it non-overlapping with `api/permission` as I thought it could be contributing to the failure somehow).

So yes, I guess what I'd like to have is
```
quarkus.keycloak.tenant.policy-enforcer.paths.1.name=Permission ResourceTenant
quarkus.keycloak.tenant.policy-enforcer.paths.1.path=/api-permission-tenant
quarkus.keycloak.tenant.policy-enforcer.paths.1.claim-information-point.claims.static-claim=static-claim
```

Can you spot what I'm missing :-) ?

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/KeycloakTestResource.java [hunk 1]
```diff
@@ -107,6 +107,7 @@ private static void configurePermissionResourcePermission(ResourceServerRepresen
                 "$evaluation.grant();\n" +
                 "}", settings);
         createPermission(settings, createResource(settings, "Permission Resource", "/api/permission"), policy);
+        createPermission(settings, createResource(settings, "Permission ResourceTenant", "/api-permission-tenant"), policy);
```
Comment (pedroigor) [inline integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/KeycloakTestResource.java:110]: Should be "Permission Resource Tenant".

File: integration-tests/keycloak-authorization/src/main/java/io/quarkus/it/keycloak/ProtectedTenantResource.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+@Path("/api-permission-tenant")
+public class ProtectedTenantResource {
+
+    @Inject
+    SecurityIdentity identity;
+
+    @GET
+    public Uni<List<Permission>> permissions() {
+        return identity.checkPermission(new AuthPermission("Permission Resource")).onItem()
```
Comment (pedroigor) [inline integration-tests/keycloak-authorization/src/main/java/io/quarkus/it/keycloak/ProtectedTenantResource.java:24]: Should be `Permission Resource Tenant`.

File: integration-tests/keycloak-authorization/src/main/resources/application.properties [hunk 3]
```diff
@@ -52,4 +53,22 @@ quarkus.keycloak.policy-enforcer.paths.8.enforcement-mode=DISABLED
...
+
+# Enable Policy Enforcement
+quarkus.keycloak.tenant.policy-enforcer.enforcement-mode=PERMISSIVE
+
+# Defines a global claim to be sent to Keycloak when evaluating permissions for any requesting coming to the application
+quarkus.keycloak.tenant.policy-enforcer.claim-information-point.claims.request-uri={request.relativePath}
+quarkus.keycloak.tenant.policy-enforcer.claim-information-point.claims.request-method={request.method}
+
+# Defines a static claim that is only sent to Keycloak when evaluating permissions for a specific path
+quarkus.keycloak.tenant.policy-enforcer.paths.1.name=Permission Resource
```
Comment (pedroigor) [inline integration-tests/keycloak-authorization/src/main/resources/application.properties:70]: Should be `Permission Resource Tenant`.

**Comment by sberyozkin** [comment]
@pedroigor thanks, let me try :-)

**Comment by sberyozkin** [comment]
@pedroigor it worked, cool, thanks

**Review by pedroigor — APPROVED** [review]

**Comment by sberyozkin** [comment]
Hey Pedro - I think we can now wire in keycloak-authorization for `web-app` if access token is configured to be a roles source, let me check it

**Pull request merged**